tensorflow/compiler/xla/service/gpu/fusion_bitcast_lift.cc

```diff
@@ -0,0 +1,205 @@
+	std::unique_ptr<HloInstruction> cloned_fusion =
+	  fusion->Clone("bitcast");
+	std::vector<HloInstruction*> stack(
+					   {cloned_fusion->fused_expression_root()});
+	bool clone_changed = false;
+	while (stack.size() > 0) {
+	  HloInstruction* i = stack.back();
+	  stack.pop_back();
+	  if (i->opcode() == HloOpcode::kTuple) {
+	    stack.insert(stack.end(), i->operands().begin(),
+			 i->operands().end());
+	    continue;
+	  } else if (i->opcode() == HloOpcode::kParameter &&
+		     absl::c_all_of(i->users(), [](HloInstruction* u) {
+                         return u->opcode() == HloOpcode::kBitcast;
+                       })) {
+	    // Replace the parameter inside the fusion.
+	    Shape new_shape = i->users()[0]->shape();
+	    int64 parameter_number = i->parameter_number();
+	    string name = i->name();
+	    auto n = HloInstruction::CreateParameter(parameter_number,
+						     new_shape, name);
+	    HloInstruction* new_parameter =
+	      i->parent()->ReplaceParameter(parameter_number,
+					    std::move(n));
+	    // Remove the old inner bitcast.
+	    auto old_users = new_parameter->users();
+	    for (HloInstruction* param_user : old_users) {
+	      DCHECK(param_user->opcode() == HloOpcode::kBitcast)
+		<< "Expected a bitcast";
+	      param_user->parent()->ReplaceInstructionWithDifferentShape(
+									 param_user, new_parameter);
+	    }
+	    // Replace the corresponding fusion operands with a new bitcast.
+	    HloInstruction* old_outer_parameter =
+	      cloned_fusion->mutable_operand(parameter_number);
+	    HloInstruction* new_op =
+	      old_outer_parameter->parent()->AddInstruction(
+                    HloInstruction::CreateBitcast(new_shape,
+						  old_outer_parameter));
+	    cloned_fusion->ReplaceOperandWithDifferentShape(
+                parameter_number, new_op);
+	    clone_changed = true;
+	    changed = true;
+	  } else if (i->opcode() == HloOpcode::kBroadcast) {
+	    // For now, do nothing. Later we can merge the broadcast
+	    // and the bitcast, but this doesn't bring benefit in my
+	    // current case.
+	    stack.push_back(i->mutable_operand(0));
+	  } else if (i->users().size() > 0 &&
+		     absl::c_all_of(i->users(), [](HloInstruction* u) {
+                         return u->opcode() == HloOpcode::kBitcast;
+                       })) {
+	    // All users are bitcast, so lift the bitcast.
+	    Shape new_shape = i->users()[0]->shape();
+	    std::vector<HloInstruction*> new_operands;
+	    for (HloInstruction* opnd : i->operands()) {
+	      Shape dtyped_new_shape = ShapeUtil::ChangeElementType(
+                  new_shape, opnd->shape().element_type());
+	      HloInstruction* new_opnd = opnd->parent()->AddInstruction(
+                  HloInstruction::CreateBitcast(dtyped_new_shape, opnd));
+	      new_operands.push_back(new_opnd);
+	      // Handle the operand right before the inserted bitcast now.
+	      if (std::find(stack.begin(), stack.end(), opnd) ==
```

Comment:
This can be pretty inefficient in the worst case. Instead, it would be better to use a absl::flat_hash_set to keep track of what is on the stack.
I will fix before merging the PR, just wanted to let you know to watch out for such patterns.

Reply:
quick reply. I do not think this is a good fix. Using `absl::flat_hash_set` would convert the stack to a set. So this would change the order of node traversal and will break the algo I think.

Reply:
Normally the fusion computation are relatively small. So this isn't a real problem until we fuse way more then right now.

Reply:
So I am not suggesting to replace the stack variable, but instead use additionally a hash set. Like in a standard DFS implementation, you would have one stack and one hash_set to mark which nodes you did already insert into the stack. So it is not a functional change, just making it more efficient.
You are right, that the fusion computations are relatively small. So maybe the original code is not as much of an issue as I first thought. But it just looks better if a DFS is implemented with linear time complexity than with quadratic time complexity.